### ISSUE TITLE
LdpPresentationGenerator fails to add public key to a VP proof node.

### ISSUE THREAD

**Issue by ricardas-buc** [eclipse-edc/IdentityHub#330]
# Bug Report

With version `0.6.2` there seem to be a change of how VP proofs are populated. Missing publicKeyJwk leads to an verification exception in LdpVerifier as it fails to correctly parse verification method.

`if (!(verificationMethod instanceof VerificationKey)) {
                    return failure("Proof did not contain a valid VerificationMethod, expected VerificationKey, got: %s".formatted(verificationMethod.getClass()));
                }`
## Possible Implementation

LdpPresentationGenerator fails to add publicKeyJwk as it was not passed to proof draft verification method. JsonWebKeyPair uses provided key to derive public key.

https://github.com/eclipse-edc/IdentityHub/blob/main/core/identity-hub-credentials/src/main/java/org/eclipse/edc/identityhub/core/creators/LdpPresentationGenerator.java#L169

Passing jwk to the proof makes VP proof valid again.
```
        var proofDraft = Jws2020ProofDraft.Builder.newInstance()
                .proofPurpose(ASSERTION_METHOD)
                .verificationMethod(new JsonWebKeyPair(URI.create(controller + "#" + publicKeyId), verificationMethodType, controllerUri, jwk))
                .created(Instant.now())
                .mapper(mapper)
                .build();
```

**Comment by paullatzelsperger** [comment]
I think you don't have the whole picture here. LD proofs can have _embedded_ JWKs, in which case you'd be correct, simply adding the `jwk` would solve the issue. But they can also be "linked", which means, you have to resolve the JWK during JSON-LD expansion. This basically downloads the JWK (which would have to be publicly available as JSON document) and inserts it into the expanded form of the signed credential. 

Generally, LDP verification has to be done on the _expanded_ form of the credential.

**As a matter of security, public key material should **not** be embedded directly, but should be referenced from the credential/presentation issuer's DID to establish provenance.**

**Comment by ricardas-buc** [comment]
> that's not a bug either :) the `DidMethodResolver` doesn't know anything about the specific verification method, JWK is just one, but there are others like RSA or Ed25519. it will simply delegate the object down to the verification infrastructure.

Sorry, I am still not fully grasping what should be correct implementation here. Is default MethodResolver not capable of passing LdpVerification checks as it will never provide correct verification method? 

To me it looks like that DataIntegrityKeyPair class implements wrong interface.

0.6.1:
This class is part of iron vc lib.
Class inheritance: DataIntegrityKeyPair -> KeyPair -> VerificationKey -> VerificationMethod

0.6.2:
Class was moved to edc:
Class inheritance: DataIntegrityKeyPair -> VerificationMethod

In both version LdpVerifier checks VerificationKey interface:

```
if (!(verificationMethod instanceof VerificationKey)) {
                return failure("%s: %s".formatted(ErrorType.Unknown, methodProperty.term()));
            }
```

**Comment by paullatzelsperger** [comment]
no, the key aspect is to remember, that all verification checks happen on the _expanded JSON-LD_ of a LDP-Credential. Expansion is the process of dereferencing and resolving all linked data, such as the public key material. 

That means, whether you embed the JWK or link it via a URL does not matter in the end, because both "variants" produce the same (ugly) JSON document. You _have_ to expand though.

The `LdpVerifier` takes a `String`, which contains the credential/presentation in JSON(-LD) format. 

If you have discovered bug, I would ask you to provide a reproducible test case, preferably as Gist, otherwise I would like to convert this issue in a discussion, as that seems more appropriate.

**Comment by ricardas-buc** [comment]
Thanks. Since its not directly identity hub issue- lets convert this one to discussion then.  I will try using referenced verification and see if it works. 

Still, LdpVerifier and DataIntegrityKeyPair code looks a bit misleading or wrong. In any case, if that part will continue blocking me - will report with a gist in connector repo :)
